lib/gauge_polyakov_loop.cu

```diff
@@ -0,0 +1,276 @@
+  class GaugeInsertTimeslice : public TunableKernel2D {
+    GaugeField &u;
+    const GaugeField &s;
+    int timeslice;
+    unsigned int minThreads() const { return s.LocalVolumeCB(); }
+
+  public:
+    GaugeInsertTimeslice(GaugeField &u, const GaugeField &s, int timeslice) :
+      TunableKernel2D(u, 2),
+      u(u),
+      s(s),
+      timeslice(timeslice)
+    {
+      strcat(aux, ",4d_vol=");
+      strcat(aux, u.VolString());
+
+      apply(device::get_default_stream());
+    }
+
+    void apply(const qudaStream_t &stream)
+    {
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+      GaugeInsertTimesliceArg<Float, nColor, recon> arg(u, s, timeslice);
+      launch<InsertTimeslice>(tp, stream, arg);
+    }
+
+    long long flops() const
+    {
+      // just a copy
+      return 0;
+    }
+
+    long long bytes() const {
+      // load timeslice, store into u
+      return 2ll * s.Bytes();
+    }
+  };
+
+  template<typename Float, int nColor, QudaReconstructType recon>
+  class GaugePolyakovLoopProduct : public TunableKernel2D {
+    GaugeField &product_field;
+    const GaugeField &u;
+    unsigned int minThreads() const { return product_field.LocalVolumeCB(); }
+
+  public:
+    GaugePolyakovLoopProduct(const GaugeField &u, GaugeField &product_field) :
+      TunableKernel2D(u, 2),
+      product_field(product_field),
+      u(u)
+    {
+      strcat(aux, ",4d_vol=");
+      strcat(aux, u.VolString());
+      strcat(aux, ",3d_vol=");
+      strcat(aux, product_field.VolString());
+
+      apply(device::get_default_stream());
+    }
+
+    void apply(const qudaStream_t &stream)
+    {
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+      GaugePolyakovLoopProductArg<Float, nColor, recon> arg(product_field, u);
+      launch<PolyakovLoopProduct>(tp, stream, arg);
+    }
+
+    long long flops() const
+    {
+      auto Nc = u.Ncolor();
+      auto mat_mul_flops = 8ll * Nc * Nc * Nc - 2 * Nc * Nc;
+      // multiplies for each loop
+      return mat_mul_flops * u.Volume() / 4;
+    }
+
+    long long bytes() const {
+      // links * one LatticeColorMatrix worth of data
+      return u.Bytes() / 4 + product_field.Bytes();
+    }
+  };
+
+  template<typename Float, int nColor, QudaReconstructType recon>
+  class GaugePolyakovLoopTrace : public TunableReduction2D {
+    const GaugeField &u;
+    using reduce_t = array<double, 2>;
+    reduce_t &ploop;
+
+  public:
+    GaugePolyakovLoopTrace(const GaugeField &u, array<double, 2> &ploop) :
+      TunableReduction2D(u),
+      u(u),
+      ploop(ploop)
+    {
+      if (u.Geometry() != QUDA_SCALAR_GEOMETRY && u.Geometry() != QUDA_VECTOR_GEOMETRY)
+        errorQuda("Invalid geometry %d in Polyakov loop calculation", u.Geometry());
+      strcat(aux, ",4d_vol=");
+      strcat(aux, u.VolString());
+      strcat(aux, ",geometry=");
+      char aux2[3];
+      u32toa(aux2, u.Geometry());
```

Comment:
And again, the geometry should already be in the `aux` key.

Reply:
Fixed locally, will push soon

Reply:
https://github.com/lattice/quda/commit/4a893d83c4c5b4c15c9e0d07f8cc3cd1a0e1955a